101-vm-user-image-data-disks/azuredeploy.json

```diff
@@ -86,9 +86,8 @@
     "subnet1Ref": "[concat(variables('vnetID'),'/subnets/',variables('subnet1Name'))]",
     "osDiskUrl": "[parameters('osDiskUrlParam')]",
     "dataDiskUrl": "[parameters('dataDiskUrlParam')]",
-    "osDiskVhdName": "[concat(variables('osDiskUrl'),'-',variables('vmName'),'.vhd')]",
-    "dataDiskVhdName": "[concat(variables('dataDiskUrl'),'-',variables('vmName'),'.vhd')]",
-    "diagStorageAccountName": "[concat(toLower(parameters('vmName')), 'diag')]"
+    "storageAccountName": "[split( split( parameters('osDiskUrlParam'), '/')[2], '.' )[0]]",
+    "diagStorageAccountName": "[concat(uniquestring(parameters('vmName')), 'diag')]"
```

Comment:
better to use something like: concat('diag', uniqueString(resourceGroup().id, 'diag')) - using only vmname as a seed will lead to collisions since vm names are not globally unique.